src/compiler/checker.ts

```diff
@@ -13353,6 +13353,17 @@ namespace ts {
                         typeParameter.constraint = getInferredTypeParameterConstraint(typeParameter) || noConstraintType;
                     }
                     else {
+                        // Detect is the constraint is for a homomorphic mapped type to a tuple and in case return a literal union of the used tuple keys
+                        if (constraintDeclaration.parent && constraintDeclaration.parent.parent && constraintDeclaration.parent.parent.kind === SyntaxKind.MappedType) {
+                            const mappedTypeNode = constraintDeclaration.parent.parent as MappedTypeNode;
+                            if (!mappedTypeNode.nameType && mappedTypeNode.typeParameter.constraint && isTypeOperatorNode(mappedTypeNode.typeParameter.constraint) && mappedTypeNode.typeParameter.constraint.operator === SyntaxKind.KeyOfKeyword) {

```

Comment:
The same condition is checked in `getTypeFromMappedTypeNode`, if I'm not mistaken. Perhaps this could be extracted to a helper function?